### DISCUSSION TITLE
Filtering attributes by cloning, swapping, and disposing is slow. Can we do better?

### DISCUSSION THREAD

**Discussion by justinmanley** [donmccurdy/glTF-Transform#520] [Q&A] [answered]
My colleagues and I are using gltf-transform to simplify lines containing in GLBs to generate levels of detail.

At each level of detail and for each line, we determine which vertices we want to retain, and which can be eliminated without introducing excessive distortion. Once this is done, we need to update the glTF so that only the desired vertices are retained. This is a very similar operation to the [`weldAndMerge`](https://github.com/donmccurdy/glTF-Transform/blob/720a7c8b66f02c6a952a1209834ba5ce9e727132/packages/functions/src/weld.ts#L58) function in `@gltf-transform/functions`, so we've adopted that approach.

For example: for each attribute, we generate a `filteredTypedArray` containing only the values corresponding to vertices we want to retain. Then, we use the following code to update it in the GLB:

```typescript
const replacementAccessor = srcAccessor.clone().setArray(filteredTypedArray);
prim.swap(srcAccessor, replacementAccessor);
if (srcAccessor.listParents().length === 1) srcAccessor.dispose();
```

We have observed that GLB processing time using this algorithm grows superlinearly with the number of primitives and accessors; a 522kb GLB with 659 primitives and 1317 accessors takes 19 seconds to process; a 21.2MB GLB with 21632 primitives and 43264 accessors takes 25 minutes to process (that's a ~30x increase in the number of primitives and a ~75x increase in processing time). As we increase the number of primitives to handle GLBs with 100s of MBs and hundreds of thousands of primitives, processing quickly becomes intractable (and often runs out of memory).

Profiling the code, I see that on the 522kb example, 11% of execution time is spent in the `onDispose()` callback triggered by `srcAccessor.dispose()`; on the 21.2MB example, 92% of execution time is spent in the same `onDispose()` callback.

It is possible to rewrite the code above as follows.
```typescript
srcAccessor.setArray(filteredTypedArray)
```

In other words: no clone, no swap, and (most importantly) no disposal.

Rewriting the code in this way reduces processing time to 14 seconds (522kb GLB) and 78 seconds (21.2MB GLB) with no observable differences in the rendered geometry.

When I make the same change to the way indices, attributes, and target attributes are handled in [`weldAndMerge`](https://github.com/donmccurdy/glTF-Transform/blob/720a7c8b66f02c6a952a1209834ba5ce9e727132/packages/functions/src/weld.ts#L58) and run the `gltf-transform` test suite, all tests (including the `weldAndMerge` test case) continue to pass.

I would like to switch our code from the `weldAndMerge` clone-swap-dispose approach to a simple `setArray`, but want to make sure doing so will not corrupt the GLB we produce.

@donmccurdy, my question for you is: why does `weldAndMerge` take the clone-swap-dispose approach to updating attribute arrays? Is this approach necessary?

**Comment by donmccurdy** [comment]
> Why does weldAndMerge take the clone-swap-dispose approach to updating attribute arrays?

The situation I'm concerned about here is (for example) one where two primitives share some — but not all — attributes. If we modify the attributes of the 1st primitive in place, the 2nd primitive may get into an invalid state. We could pre-process the scene to make copies of any shared attributes without changing the result — the current code separates any shared attributes later in the process anyway. I haven't looked closely, but that (along with the change you suggest above) might get us to `O(n)` time here?

Alternatively - the internal [Graph](https://github.com/donmccurdy/property-graph/blob/main/src/graph.ts) class used for most of these clone/dispose/swap operations is pretty simple, but not really optimized for heavy throughput. Most of my work is with models destined for WebGL applications, where performance targets are usually closer to 100-1000 primitives. Perhaps some optimizations in the Graph class would allow `dispose()` to run much more quickly.

**Reply by justinmanley** [reply]
Thanks for the super-fast response!

What you say about multiple primitives pointing to the same attribute makes total sense. One case where we would absolutely need to split the attributes is if multiple primitives with different modes (say, one `LINES` and one `LINE_STRIP`) used the same position attribute. Because of the different drawing modes, the primitives would actually represent different geometries despite sharing the same position attribute, and so simplifying them would produce different results. I can see that this is a case where we would need to split the attributes in order to get a correct result. Thanks for pointing this out -- I hadn't thought about this case before.

Thinking about your suggestion for optimizations in the `Graph` class which might make `dispose()` run more quickly...

After digging into the call profile of the `onDispose()` callback, I can confirm that disposing all attributes in the graph is quadratic in the number of attributes. This is because every attribute has a link (I say "link" because we're still using `v1.2.4`) to the `Root`, and so every time an attribute node is `dispose`d, the `onDispose` callback [must iterate through every single attribute](https://github.com/donmccurdy/glTF-Transform/blob/78cbc369b2808c2840a926c3ff57f397a6eb3f70/packages/core/src/graph/graph-node.ts#L77) on the root in order to find the correct `Root -> Attribute` link to dispose, and then iterate through every single attribute again (minus 1) to add them back to the cleared array. I see that the [variable names are different in the latest version, but the logic is still the same](https://github.com/donmccurdy/property-graph/blob/main/src/graph-node.ts#L246).

We can make disposal of a single attribute constant-time rather than linear-time if we use a data structure which allows constant-time lookup and removal of links/refs. **What would you think of making the collection of refs a `Set` rather than an `Array`?** Then the relevant code (in the latest version of the `property-graph` package) would look like:

```typescript
protected addRef<K extends RefListKeys<Attributes>>(
    attribute: K,
    value: GraphNode & Attributes[K][keyof Attributes[K]],
    attributes?: Record<string, unknown>
): this {
    const ref = this.graph.createEdge(attribute as string, this, value, attributes);

    const refs = this[$attributes][attribute] as Set<Ref>;
    refs.add(ref);

    ref.addEventListener('dispose', () => {
      refs.delete(ref); // constant-time!
      this.dispatchEvent({ type: 'change', attribute });
    });

    return this.dispatchEvent({ type: 'change', attribute });
}
```

**Reply by donmccurdy** [reply]
> What would you think of making the collection of refs a Set rather than an Array? 

Good point! That's probably the right direction (just filed as https://github.com/donmccurdy/property-graph/issues/20). Making the change in a way that is backward-compatible with the extension API will be a little difficult, and I'm hesitant to make breaking changes so soon after v2. It can be done but may not happen just yet.

If you know your attributes won't be shared — or can detach those that are — then I think it's safe to modify attributes in place and work around the performance bottleneck that way for now.

**Reply by justinmanley** [reply]
sounds good, thanks!

I'm moving ahead with the workaround you suggested. Added some details as an answer.

thanks again!

**Comment by justinmanley** [comment, marked as answer]
I was able to significantly improve performance using the workaround suggested by @donmccurdy (thanks!). With this workaround, processing time is reduced from 25 minutes for a 22MB mesh to 50 seconds.

Here's the workaround:
- If an accessor is used by multiple primitives, clone it and swap the original and the clone.  The swap operation (like dispose) is asymptotically inefficient, since it requires iterating over all the (accessor) children of the current primitive, but because primitives typically have only a small bounded number of accessors (e.g. one each for indices, position, and color), this inefficiency impairs performance only by a small constant factor.
- If an accessor is used only by a single primitive, modify it directly in place (this reuse eliminates the need for expensive disposal)

In code:

```typescript 
// Helper
function isUsedMultipleTimes(accessor: Accessor): boolean {
  // Inefficient, but doesn't matter as long as accessors have only a small number of parents. 
  // In all the examples I've tested in practice, accessors have only two parents: the root, and 
  // a single primitive.
  const nonRootParents = accessor
    .listParents()
    .filter(parent => parent.propertyType !== 'Root');
  return nonRootParents.length > 1;
}
```

Accessors:
```typescript
// newValues is a TypedArray
if (isUsedMultipleTimes(srcAccessor)) {
  const replacementAccessor = srcAccessor
    .clone()
    .setArray(newValues);
  primitive.swap(srcAccessor, replacementAccessor);
} else {
  srcAccessor.setArray(newValues);
}
```

Indices:
```typescript
if (isUsedMultipleTimes(srcIndices)) {
  primitive.setIndices(srcIndices.clone().setArray(replacementIndicesArray));
} else {
  srcIndices.setArray(replacementIndicesArray);
}
```
